File: Leetcode/PY/Leetcode525-连续数组.py
from typing import List
from collections import defaultdict
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        ans,mp,cur = 0,defaultdict(int),0
        n = len(nums)
        cnt = 0
        mp[0] = -1
        for i in range(n):
            cur = 1 if nums[i] == 1 else -1
            cnt += cur
            if cnt in mp:
                ans = max(ans,i - mp[cnt])
            else:
                mp[cnt] = i
        return ans

File: Leetcode/Rust/Leetcode525-连续数组.rs
use std::collections::HashMap;
impl Solution {
    pub fn find_max_length(nums: Vec<i32>) -> i32 {
        let (mut ans,n,mut cur) = (0,nums.len(),0);
        let mut mp = HashMap::new();
        let mut cnt = 0;
        mp.insert(0, -1);
        for (idx,num) in nums.into_iter().enumerate(){
            match num {
                1 =>{
                    cur = 1;
                },
                _ =>{
                    cur = -1;
                }
            }
            cnt += cur;
            if mp.contains_key(&cnt){
                ans = ans.max(idx as i32 - mp[&cnt]);
            }else{
                mp.entry(cnt).or_insert(idx as i32);
            }
        }
        ans
    }
}